urlb57eb64056_jacadenac_ISI_Othello_UN_tgz-pJ8-othello_OthelloJ8/src/logica/Controlador.java

```diff
@@ -0,0 +1,375 @@
+		Point punto;
+		if (this.turno == Turno.NEGRAS) {
+			funcionEvaluacion = new EvaluacionEstrategica(jugadorNegro.getEstrategia());
+			punto = buscador.busquedaSimple(tablero, tablero.obtenerTipoTablero(), jugadorNegro.color(), 
+					profundidad, funcionEvaluacion).obtenerPunto();
+		} else {
+			funcionEvaluacion = new EvaluacionEstrategica(jugadorBlanco.getEstrategia());
+			punto = buscador.busquedaSimple(tablero, tablero.obtenerTipoTablero(), jugadorBlanco.color(), 
+					profundidad, funcionEvaluacion).obtenerPunto();
+		}
+		return punto;
+	}
+	
+	public Set<Point> marcarMovidasPosibles() {
+		Set<Point> movidas;
+		if (this.turno == Turno.NEGRAS) {
+			movidas = tablero.obtenerMovidasPosibles(jugadorNegro);
+		} else {
+			movidas = tablero.obtenerMovidasPosibles(jugadorBlanco);
+		}
+		tablero.marcarMovidasPosibles(movidas);
+		puedeMover = movidas.isEmpty() ? ++puedeMover : PUEDEMOVER;
+		return movidas;
+	}
+	
+	public void desmarcarMovidasPosibles() {
+		tablero.desmarcarMovidasPosibles();
+	}
+	
+	public void actualizarEstadoCasilla(Point punto, EstadoCasilla estadoCasilla) {
+		tablero.actualizarEstadoCasilla(punto, estadoCasilla);
+	}
+
+	public int obtenerPuntajeNegro() {
+		jugadorNegro.setPuntaje(tablero.contar(EstadoCasilla.BLACK));
+		return jugadorNegro.getPuntaje();
+	}
+	
+	public int obtenerPuntajeBlanco() {
+		jugadorBlanco.setPuntaje(tablero.contar(EstadoCasilla.WHITE));
+		return jugadorBlanco.getPuntaje();
+	}
+	
+	public Jugador obtenerGanador() {
+		return jugadorNegro.getPuntaje() < jugadorBlanco.getPuntaje() ? jugadorBlanco : jugadorNegro;
+	}
+	
+	public boolean esEmpate() {
+		return jugadorNegro.getPuntaje() == jugadorBlanco.getPuntaje();
+	}
+	
+	public boolean finDelJuego() {	
+		if(tablero.estaLleno() || verificarPuntajeEnCero() || puedeMover == NOPUEDEMOVER){
+			juegoTerminado = true;
+		}
+		return juegoTerminado;
+	}
+	
+	private boolean verificarPuntajeEnCero() {
+		return obtenerPuntajeNegro() == 0 || obtenerPuntajeBlanco() == 0;
+	}
+	
+	public void cambiarTurno() {
+		if (turno ==  Turno.BLANCAS) {
+			turno = Turno.NEGRAS;
+		} else {
+			turno = Turno.BLANCAS;
+		}
+	}
+	
+	public Jugador jugadorActual() {
+		if (turno ==  Turno.BLANCAS) {
+			return jugadorBlanco;
+		} else {
+			return jugadorNegro;
+		}
+	}
+
+	public Tablero obtenerTableroLogica(){
+		return tablero;
+	}
+	
+	private static class soporteControlador {
+		private static final Controlador INSTANCE = new Controlador();
+	}
+	
+	public static Controlador obtenerInstancia() {
+		return soporteControlador.INSTANCE;
+	}
+	
+	private static int[][] leerEstrategias() {
+	    int[][] estrategias = new int[2][4];
+		JFileChooser fileChooser = new JFileChooser();
+		String nombreFichero = "dosMejores.txt";
+		File fichero = new File(directorio.getPath(),nombreFichero);
+		fileChooser.setCurrentDirectory(fichero);		
+		try {
+			BufferedReader readerPenultimaLinea = new BufferedReader(new FileReader(fichero));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *readerPenultimaLinea*. The resource is referenced at lines 238, 246. The resource is closed at line 256. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *readerPenultimaLinea* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).